tests/database/test_journal_db.py

```diff
@@ -47,30 +63,137 @@ def test_revert_clears_reverted_journal_entries(journal_db):
 
     assert journal_db.get(b'1') == b'test-a'
 
-    snapshot_a = journal_db.snapshot()
+    changeset_a = journal_db.record()
 
     journal_db.set(b'1', b'test-b')
     journal_db.delete(b'1')
     journal_db.set(b'1', b'test-c')
 
     assert journal_db.get(b'1') == b'test-c'
 
-    snapshot_b = journal_db.snapshot()
+    changeset_b = journal_db.record()
 
     journal_db.set(b'1', b'test-d')
     journal_db.delete(b'1')
     journal_db.set(b'1', b'test-e')
 
     assert journal_db.get(b'1') == b'test-e'
 
-    journal_db.revert(snapshot_b)
+    journal_db.discard(changeset_b)
 
     assert journal_db.get(b'1') == b'test-c'
 
     journal_db.delete(b'1')
 
     assert journal_db.exists(b'1') is False
 
-    journal_db.revert(snapshot_a)
+    journal_db.discard(changeset_a)
+
+    assert journal_db.get(b'1') == b'test-a'
+
+
+def test_revert_removes_journal_entries(journal_db):
+
+    changeset_a = journal_db.record()  # noqa: F841
+    assert len(journal_db.journal.journal_data) == 2
+
+    changeset_b = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 3
+
+    # Forget *latest* changeset and prove it's the only one removed
+    journal_db.discard(changeset_b)
+    assert len(journal_db.journal.journal_data) == 2
+
+    changeset_b2 = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 3
+
+    changeset_c = journal_db.record()  # noqa: F841
+    assert len(journal_db.journal.journal_data) == 4
+
+    changeset_d = journal_db.record()  # noqa: F841
+    assert len(journal_db.journal.journal_data) == 5
+
+    # Forget everything from b2 (inclusive) and what follows
+    journal_db.discard(changeset_b2)
+    assert len(journal_db.journal.journal_data) == 2
+    assert journal_db.journal.has_changeset(changeset_b2) is False
+
+
+def test_commit_merges_changeset_into_previous(journal_db):
+
+    changeset = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 2
+
+    journal_db.set(b'1', b'test-a')
+    assert journal_db.get(b'1') == b'test-a'
+
+    journal_db.commit(changeset)
+
+    assert len(journal_db.journal.journal_data) == 1
+    assert journal_db.journal.has_changeset(changeset) is False
+
+
+def test_committing_middle_changeset_merges_in_subsequent_changesets(journal_db):
+
+    journal_db.set(b'1', b'test-a')
+    changeset_a = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 2
+
+    journal_db.set(b'1', b'test-b')
+    changeset_b = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 3
+
+    journal_db.set(b'1', b'test-c')
+    changeset_c = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 4
+
+    # TODO: Clarify
+    # This seems counterintuitive to me. Why does commiting to changeset_b
+    # Mean we are implicitly commiting to subsequent changesets?
+    # Why don't we just merge b into a but keep c seperate?
```

Comment:
This comment should either be rewritten or removed assuming you now understand why this happens.  Probably worth including such an explanitory comment in the actual `commit()` or underlying `Journal` method since it's not intuitive functionality unless you know how the journal is intended to be used.

Reply:
I removed the (left over) comment here. The doc comment of `commit()` already cover that to some extend. I think further elaboration there would make that API comment rather wordy. Maybe there's a better place in the upcoming docs for a deep dive into the internals of the `JournalDB`